### ISSUE TITLE
list comprehension with None (falsy) values filtered by condition

### ISSUE THREAD

**Issue by pawamoy** [astral-sh/ty#680]
### Summary

```python
import os


def func(param: str) -> str:
    return param


l = [func(val) for var in ("HELLO", "WORLD") if (val := os.getenv(var))]
```

```console
% uvx ty check t.py
WARN ty is pre-release software and not ready for production use. Expect to encounter bugs, missing features, and fatal errors.
error[invalid-argument-type]: Argument to function `func` is incorrect
 --> t.py:8:11
  |
8 | l = [func(val) for var in ("HELLO", "WORLD") if (val := os.getenv(var))]
  |           ^^^ Expected `str`, found `str | None`
  |
info: Function defined here
 --> t.py:4:5
  |
4 | def func(param: str) -> str:
  |     ^^^^ ---------- Parameter declared here
5 |     return param
  |
info: rule `invalid-argument-type` is enabled by default

Found 1 diagnostic
```

Here `func` cannot ever be called with `None` since such falsy values are filtered by the list comprehension condition `if (val := os.getenv(var))`.

### Version

ty 0.0.1-alpha.11

**Comment by AlexWaygood** [comment]
haven't looked in depth but this might be the same underlying issue as #626

**Comment by sharkdp** [comment]
> haven't looked in depth but this might be the same underlying issue as [#626](https://github.com/astral-sh/ty/issues/626)

I don't think so? There's no boolean operator here, and something like the following works just fine:
```py
if val := os.getenv("HELLO"):
    func(val)
```

It looks like we don't record narrowing constraints for generator expressions at all ([ref](https://github.com/astral-sh/ruff/blob/e1809752269c65418f396671afac042bec3b7a86/crates/ty_python_semantic/src/semantic_index/builder.rs#L857-L859))?
